Sources/SwiftGRPCNIO/ClientCalls/BaseClientCall.swift

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Foundation
+import NIO
+import NIOHTTP1
+import NIOHTTP2
+import SwiftProtobuf
+
+public class BaseClientCall<RequestMessage: Message, ResponseMessage: Message> {
+  /// The underlying `GRPCClient` providing the HTTP/2 channel and multiplexer.
+  internal let client: GRPCClient
+
+  /// Promise for an HTTP/2 stream.
+  internal let streamPromise: EventLoopPromise<Channel>
+
+  /// Client channel handler. Handles some internal state for reading/writing messages to the channel.
+  /// The handler also owns the promises for the futures that this class surfaces to the user (such as
+  /// `initialMetadata` and `status`).
+  internal let clientChannelHandler: GRPCClientChannelHandler<RequestMessage, ResponseMessage>
+
+  /// Sets up a gRPC call.
+  ///
+  /// A number of actions are performed:
+  /// - a new HTTP/2 stream is created and configured using the channel and multiplexer provided by `client`,
+  /// - a callback is registered on the new stream (`subchannel`) to send the request head,
+  /// - a timeout is scheduled if one is set in the `callOptions`.
```

Comment:
Absolutely. It would be good to have some kind of tracking for planned NIO work -- I saw @timburks put a couple of cards into the projects section. Or we could just maintain an issue with a checklist of outstanding work.

Reply:
Agreed. For now I'd just suggest having a checklist issue.